// @oakserver/oak/script/etag.js
"use strict";
// Copyright 2018-2023 the oak authors. All rights reserved. MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.factory = exports.getEntity = exports.ifNoneMatch = exports.ifMatch = exports.calculate = void 0;
const tslib_1 = require("tslib");
/**
 * A collection of oak specific APIs for management of ETags.
 *
 * @module
 */
const dntShim = tslib_1.__importStar(require("./_dnt.shims.js"));
const deps_js_1 = require("./deps.js");
const util_js_1 = require("./util.js");
// re-exports to maintain backwards compatibility
var deps_js_2 = require("./deps.js");
Object.defineProperty(exports, "calculate", { enumerable: true, get: function () { return deps_js_2.calculate; } });
Object.defineProperty(exports, "ifMatch", { enumerable: true, get: function () { return deps_js_2.ifMatch; } });
Object.defineProperty(exports, "ifNoneMatch", { enumerable: true, get: function () { return deps_js_2.ifNoneMatch; } });
function fstat(file) {
    if ("fstat" in dntShim.Deno) {
        // deno-lint-ignore no-explicit-any
        return dntShim.Deno.fstat(file.rid);
    }
    return Promise.resolve(undefined);
}
/** For a given Context, try to determine the response body entity that an ETag
 * can be calculated from. */
// deno-lint-ignore no-explicit-any
function getEntity(context) {
    const { body } = context.response;
    if (body instanceof dntShim.Deno.FsFile) {
        return fstat(body);
    }
    if (body instanceof Uint8Array) {
        return Promise.resolve(body);
    }
    if (util_js_1.BODY_TYPES.includes(typeof body)) {
        return Promise.resolve(String(body));
    }
    if ((0, util_js_1.isAsyncIterable)(body) || (0, util_js_1.isReader)(body)) {
        return Promise.resolve(undefined);
    }
    if (typeof body === "object" && body !== null) {
        try {
            const bodyText = JSON.stringify(body);
            return Promise.resolve(bodyText);
        }
        catch {
            // We don't really care about errors here
        }
    }
    return Promise.resolve(undefined);
}
exports.getEntity = getEntity;
/**
 * Create middleware that will attempt to decode the response.body into
 * something that can be used to generate an `ETag` and add the `ETag` header to
 * the response.
 */
// deno-lint-ignore no-explicit-any
function factory(options) {
    return async function etag(context, next) {
        await next();
        if (!context.response.headers.has("ETag")) {
            const entity = await getEntity(context);
            if (entity) {
                const etag = await (0, deps_js_1.calculate)(entity, options);
                if (etag) {
                    context.response.headers.set("ETag", etag);
                }
            }
        }
    };
}
exports.factory = factory;
