// @oakserver/oak/script/node_shims.js
"use strict";
// Copyright 2018-2023 the oak authors. All rights reserved. MIT license.
var _ErrorEvent_message, _ErrorEvent_filename, _ErrorEvent_lineno, _ErrorEvent_colno, _ErrorEvent_error;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorEvent = void 0;
const tslib_1 = require("tslib");
class ErrorEvent extends Event {
    get message() {
        return tslib_1.__classPrivateFieldGet(this, _ErrorEvent_message, "f");
    }
    get filename() {
        return tslib_1.__classPrivateFieldGet(this, _ErrorEvent_filename, "f");
    }
    get lineno() {
        return tslib_1.__classPrivateFieldGet(this, _ErrorEvent_lineno, "f");
    }
    get colno() {
        return tslib_1.__classPrivateFieldGet(this, _ErrorEvent_colno, "f");
    }
    // deno-lint-ignore no-explicit-any
    get error() {
        return tslib_1.__classPrivateFieldGet(this, _ErrorEvent_error, "f");
    }
    constructor(type, eventInitDict = {}) {
        super(type, eventInitDict);
        _ErrorEvent_message.set(this, void 0);
        _ErrorEvent_filename.set(this, void 0);
        _ErrorEvent_lineno.set(this, void 0);
        _ErrorEvent_colno.set(this, void 0);
        // deno-lint-ignore no-explicit-any
        _ErrorEvent_error.set(this, void 0);
        const { message = "error", filename = "", lineno = 0, colno = 0, error } = eventInitDict;
        tslib_1.__classPrivateFieldSet(this, _ErrorEvent_message, message, "f");
        tslib_1.__classPrivateFieldSet(this, _ErrorEvent_filename, filename, "f");
        tslib_1.__classPrivateFieldSet(this, _ErrorEvent_lineno, lineno, "f");
        tslib_1.__classPrivateFieldSet(this, _ErrorEvent_colno, colno, "f");
        tslib_1.__classPrivateFieldSet(this, _ErrorEvent_error, error, "f");
    }
}
exports.ErrorEvent = ErrorEvent;
_ErrorEvent_message = new WeakMap(), _ErrorEvent_filename = new WeakMap(), _ErrorEvent_lineno = new WeakMap(), _ErrorEvent_colno = new WeakMap(), _ErrorEvent_error = new WeakMap();
