// @oakserver/oak/script/testing.js
"use strict";
// Copyright 2018-2023 the oak authors. All rights reserved. MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockNext = exports.createMockContext = exports.mockContextState = exports.createMockApp = void 0;
const tslib_1 = require("tslib");
// deno-lint-ignore-file no-explicit-any
/**
 * A collection of utility APIs which can make testing of an oak application
 * easier.
 *
 * @module
 */
const dntShim = tslib_1.__importStar(require("./_dnt.shims.js"));
const deps_js_1 = require("./deps.js");
const response_js_1 = require("./response.js");
/** Creates a mock of `Application`. */
function createMockApp(state = {}) {
    const app = {
        state,
        use() {
            return app;
        },
        [Symbol.for("Deno.customInspect")]() {
            return "MockApplication {}";
        },
        [Symbol.for("nodejs.util.inspect.custom")](depth, options, inspect) {
            if (depth < 0) {
                return options.stylize(`[MockApplication]`, "special");
            }
            const newOptions = Object.assign({}, options, {
                depth: options.depth === null ? null : options.depth - 1,
            });
            return `${options.stylize("MockApplication", "special")} ${inspect({}, newOptions)}`;
        },
    };
    return app;
}
exports.createMockApp = createMockApp;
/** Allows external parties to modify the context state. */
exports.mockContextState = {
    /** Adjusts the return value of the `acceptedEncodings` in the context's
     * `request` object. */
    encodingsAccepted: "identity",
};
/** Create a mock of `Context` or `RouterContext`. */
function createMockContext({ ip = "127.0.0.1", method = "GET", params, path = "/", state, app = createMockApp(state), headers: requestHeaders, } = {}) {
    function createMockRequest() {
        const headers = new dntShim.Headers(requestHeaders);
        return {
            accepts(...types) {
                if (!headers.has("Accept")) {
                    return;
                }
                if (types.length) {
                    return (0, deps_js_1.accepts)({ headers }, ...types);
                }
                return (0, deps_js_1.accepts)({ headers });
            },
            acceptsEncodings() {
                return exports.mockContextState.encodingsAccepted;
            },
            headers,
            ip,
            method,
            path,
            search: undefined,
            searchParams: new URLSearchParams(),
            url: new URL(path, "http://localhost/"),
        };
    }
    const request = createMockRequest();
    const response = new response_js_1.Response(request);
    const cookies = new deps_js_1.SecureCookieMap(request, { response });
    return {
        app,
        params,
        request,
        cookies,
        response,
        state: Object.assign({}, app.state),
        assert(condition, errorStatus = 500, message, props) {
            if (condition) {
                return;
            }
            const err = (0, deps_js_1.createHttpError)(errorStatus, message);
            if (props) {
                Object.assign(err, props);
            }
            throw err;
        },
        throw(errorStatus, message, props) {
            const err = (0, deps_js_1.createHttpError)(errorStatus, message);
            if (props) {
                Object.assign(err, props);
            }
            throw err;
        },
        [Symbol.for("Deno.customInspect")]() {
            return `MockContext {}`;
        },
        [Symbol.for("nodejs.util.inspect.custom")](depth, options, inspect) {
            if (depth < 0) {
                return options.stylize(`[MockContext]`, "special");
            }
            const newOptions = Object.assign({}, options, {
                depth: options.depth === null ? null : options.depth - 1,
            });
            return `${options.stylize("MockContext", "special")} ${inspect({}, newOptions)}`;
        },
    };
}
exports.createMockContext = createMockContext;
/** Creates a mock `next()` function which can be used when calling
 * middleware. */
function createMockNext() {
    return async function next() { };
}
exports.createMockNext = createMockNext;
