// windmill-client/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.usernameToEmail = exports.getFlowUserState = exports.setFlowUserState = exports.getRootJobId = exports.waitJob = exports.runScriptAsync = exports.runScript = exports.task = exports.writeS3File = exports.loadS3File = exports.loadS3FileStream = exports.denoS3LightClientSettings = exports.getIdToken = exports.getState = exports.setState = exports.getResumeUrls = exports.setResource = exports.getResource = exports.setVariable = exports.getVariable = exports.setClient = exports.OpenAPI = exports.CancelError = exports.CancelablePromise = exports.ApiError = void 0;
// This file is auto-generated by @hey-api/openapi-ts
var ApiError_1 = require("./core/ApiError");
Object.defineProperty(exports, "ApiError", { enumerable: true, get: function () { return ApiError_1.ApiError; } });
var CancelablePromise_1 = require("./core/CancelablePromise");
Object.defineProperty(exports, "CancelablePromise", { enumerable: true, get: function () { return CancelablePromise_1.CancelablePromise; } });
Object.defineProperty(exports, "CancelError", { enumerable: true, get: function () { return CancelablePromise_1.CancelError; } });
var OpenAPI_1 = require("./core/OpenAPI");
Object.defineProperty(exports, "OpenAPI", { enumerable: true, get: function () { return OpenAPI_1.OpenAPI; } });
__exportStar(require("./services.gen"), exports);
__exportStar(require("./types.gen"), exports);
var client_1 = require("./client");
Object.defineProperty(exports, "setClient", { enumerable: true, get: function () { return client_1.setClient; } });
Object.defineProperty(exports, "getVariable", { enumerable: true, get: function () { return client_1.getVariable; } });
Object.defineProperty(exports, "setVariable", { enumerable: true, get: function () { return client_1.setVariable; } });
Object.defineProperty(exports, "getResource", { enumerable: true, get: function () { return client_1.getResource; } });
Object.defineProperty(exports, "setResource", { enumerable: true, get: function () { return client_1.setResource; } });
Object.defineProperty(exports, "getResumeUrls", { enumerable: true, get: function () { return client_1.getResumeUrls; } });
Object.defineProperty(exports, "setState", { enumerable: true, get: function () { return client_1.setState; } });
Object.defineProperty(exports, "getState", { enumerable: true, get: function () { return client_1.getState; } });
Object.defineProperty(exports, "getIdToken", { enumerable: true, get: function () { return client_1.getIdToken; } });
Object.defineProperty(exports, "denoS3LightClientSettings", { enumerable: true, get: function () { return client_1.denoS3LightClientSettings; } });
Object.defineProperty(exports, "loadS3FileStream", { enumerable: true, get: function () { return client_1.loadS3FileStream; } });
Object.defineProperty(exports, "loadS3File", { enumerable: true, get: function () { return client_1.loadS3File; } });
Object.defineProperty(exports, "writeS3File", { enumerable: true, get: function () { return client_1.writeS3File; } });
Object.defineProperty(exports, "task", { enumerable: true, get: function () { return client_1.task; } });
Object.defineProperty(exports, "runScript", { enumerable: true, get: function () { return client_1.runScript; } });
Object.defineProperty(exports, "runScriptAsync", { enumerable: true, get: function () { return client_1.runScriptAsync; } });
Object.defineProperty(exports, "waitJob", { enumerable: true, get: function () { return client_1.waitJob; } });
Object.defineProperty(exports, "getRootJobId", { enumerable: true, get: function () { return client_1.getRootJobId; } });
Object.defineProperty(exports, "setFlowUserState", { enumerable: true, get: function () { return client_1.setFlowUserState; } });
Object.defineProperty(exports, "getFlowUserState", { enumerable: true, get: function () { return client_1.getFlowUserState; } });
Object.defineProperty(exports, "usernameToEmail", { enumerable: true, get: function () { return client_1.usernameToEmail; } });
