// @oakserver/oak/esm/range.js
var _MultiPartStream_contentLength, _MultiPartStream_postscript, _MultiPartStream_preamble;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
// Copyright 2018-2023 the oak authors. All rights reserved. MIT license.
import * as dntShim from "./_dnt.shims.js";
import { concat, contentType, copyBytes, createHttpError, Status, } from "./deps.js";
import { calculate } from "./etag.js";
import { assert, DEFAULT_CHUNK_SIZE } from "./util.js";
const ETAG_RE = /(?:W\/)?"[ !#-\x7E\x80-\xFF]+"/;
/** Determine, by the value of an `If-Range` header, if a `Range` header should
 * be applied to a request, returning `true` if it should or otherwise
 * `false`. */
export async function ifRange(value, mtime, entity) {
    if (value) {
        const matches = value.match(ETAG_RE);
        if (matches) {
            const [match] = matches;
            if (await calculate(entity) === match) {
                return true;
            }
        }
        else {
            return new Date(value).getTime() >= mtime;
        }
    }
    return false;
}
export function parseRange(value, size) {
    const ranges = [];
    const [unit, rangesStr] = value.split("=");
    if (unit !== "bytes") {
        throw createHttpError(Status.RequestedRangeNotSatisfiable);
    }
    for (const range of rangesStr.split(/\s*,\s+/)) {
        const item = range.split("-");
        if (item.length !== 2) {
            throw createHttpError(Status.RequestedRangeNotSatisfiable);
        }
        const [startStr, endStr] = item;
        let start;
        let end;
        try {
            if (startStr === "") {
                start = size - parseInt(endStr, 10) - 1;
                end = size - 1;
            }
            else if (endStr === "") {
                start = parseInt(startStr, 10);
                end = size - 1;
            }
            else {
                start = parseInt(startStr, 10);
                end = parseInt(endStr, 10);
            }
        }
        catch {
            throw createHttpError();
        }
        if (start < 0 || start >= size || end < 0 || end >= size || start > end) {
            throw createHttpError(Status.RequestedRangeNotSatisfiable);
        }
        ranges.push({ start, end });
    }
    return ranges;
}
/** A reader  */
async function readRange(file, range) {
    let length = range.end - range.start + 1;
    assert(length);
    await file.seek(range.start, dntShim.Deno.SeekMode.Start);
    const result = new Uint8Array(length);
    let off = 0;
    while (length) {
        const p = new Uint8Array(Math.min(length, DEFAULT_CHUNK_SIZE));
        const nread = await file.read(p);
        assert(nread !== null, "Unexpected EOF encountered when reading a range.");
        assert(nread > 0, "Unexpected read of 0 bytes while reading a range.");
        copyBytes(p, result, off);
        off += nread;
        length -= nread;
        assert(length >= 0, "Unexpected length remaining.");
    }
    return result;
}
const encoder = new TextEncoder();
/** A class that takes a file (either a Deno.FsFile or Uint8Array) and bytes
 * and streams the ranges as a multi-part encoded HTTP body.
 *
 * This is specifically used by the `.send()` functionality to fulfill range
 * requests it receives, and could be used by others when trying to deal with
 * range requests, but is generally a low level API that most users of oak
 * would not need to worry about. */
export class MultiPartStream extends dntShim.ReadableStream {
    constructor(file, type, ranges, size, boundary) {
        super({
            pull: async (controller) => {
                const range = ranges.shift();
                if (!range) {
                    controller.enqueue(__classPrivateFieldGet(this, _MultiPartStream_postscript, "f"));
                    controller.close();
                    if (!(file instanceof Uint8Array)) {
                        file.close();
                    }
                    return;
                }
                let bytes;
                if (file instanceof Uint8Array) {
                    bytes = file.subarray(range.start, range.end + 1);
                }
                else {
                    bytes = await readRange(file, range);
                }
                const rangeHeader = encoder.encode(`Content-Range: ${range.start}-${range.end}/${size}\n\n`);
                controller.enqueue(concat([__classPrivateFieldGet(this, _MultiPartStream_preamble, "f"), rangeHeader, bytes]));
            },
        });
        _MultiPartStream_contentLength.set(this, void 0);
        _MultiPartStream_postscript.set(this, void 0);
        _MultiPartStream_preamble.set(this, void 0);
        const resolvedType = contentType(type);
        if (!resolvedType) {
            throw new TypeError(`Could not resolve media type for "${type}"`);
        }
        __classPrivateFieldSet(this, _MultiPartStream_preamble, encoder.encode(`\n--${boundary}\nContent-Type: ${resolvedType}\n`), "f");
        __classPrivateFieldSet(this, _MultiPartStream_postscript, encoder.encode(`\n--${boundary}--\n`), "f");
        __classPrivateFieldSet(this, _MultiPartStream_contentLength, ranges.reduce((prev, { start, end }) => {
            return prev + __classPrivateFieldGet(this, _MultiPartStream_preamble, "f").length + String(start).length +
                String(end).length + String(size).length + 20 + (end - start);
        }, __classPrivateFieldGet(this, _MultiPartStream_postscript, "f").length), "f");
    }
    /** The content length of the entire streamed body. */
    contentLength() {
        return __classPrivateFieldGet(this, _MultiPartStream_contentLength, "f");
    }
}
_MultiPartStream_contentLength = new WeakMap(), _MultiPartStream_postscript = new WeakMap(), _MultiPartStream_preamble = new WeakMap();
