// @ayonli/jsext/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _try = require('./try.js');
var func = require('./func.js');
var wrap = require('./wrap.js');
var mixin = require('./mixin.js');
var throttle = require('./throttle.js');
var debounce = require('./debounce.js');
var queue = require('./queue.js');
var lock = require('./lock.js');
var reader = require('./reader.js');
var chan = require('./chan.js');
var parallel = require('./parallel.js');
var run = require('./run.js');
var example$1 = require('./example.js');
var deprecate = require('./deprecate.js');
var pipe = require('./pipe.js');
var _class = require('./class.js');
var types = require('./types.js');
var reader_util = require('./reader/util.js');

/** @deprecated import `Queue` from `@ayonli/jsext/queue` instead. */
const Queue = queue.Queue;
/** @deprecated import `Mutex` from `@ayonli/jsext/lock` instead. */
const Mutex = lock.Mutex;
/** @deprecated import `Channel` from `@ayonli/jsext/chan` instead. */
const Channel = chan.Channel;
/**
 * The entry of jsext major functions.
 */
const jsext = {
    _try: _try.default,
    /** @deprecated use `_try` instead */
    try: _try.default,
    func: func.default,
    wrap: wrap.default,
    mixin: mixin.default,
    throttle: throttle.default,
    debounce: debounce.default,
    queue: queue.default,
    lock: lock.default,
    /** @deprecated Use {@link toAsyncIterable} from `@ayonli/jsext/reader` instead. */
    read: reader_util.toAsyncIterable,
    /** @deprecated Use {@link readAsArray} from `@ayonli/jsext/reader` instead. */
    readAll: reader.readAsArray,
    chan: chan.default,
    parallel: parallel.default,
    run: run.default,
    /** @deprecated */
    example: example$1.default,
    deprecate: deprecate.default,
    pipe: pipe.default,
    /** @deprecated import `isClass` from `@ayonli/jsext/class` instead. */
    isClass: _class.isClass,
    /** @deprecated import `isSubclassOf` from `@ayonli/jsext/class` instead. */
    isSubclassOf: _class.isSubclassOf,
    /** @deprecated use `mixin` instead */
    mixins: mixin.default,
};
/** @deprecated Use {@link toAsyncIterable} from `@ayonli/jsext/reader` instead. */
const read = reader_util.toAsyncIterable;
/** @deprecated Use {@link readAsArray} from `@ayonli/jsext/reader` instead. */
const readAll = reader.readAsArray;
/** @deprecated */
const example = example$1.default;
/** @deprecated import `isClass` from `@ayonli/jsext/class` instead. */
const isClass = _class.isClass;
/** @deprecated import `isSubclassOf` from `@ayonli/jsext/class` instead. */
const isSubclassOf = _class.isSubclassOf;
/** @deprecated use `mixin` instead */
const mixins = mixin.default;

exports._try = _try.default;
exports.func = func.default;
exports.wrap = wrap.default;
exports.mixin = mixin.default;
exports.throttle = throttle.default;
exports.debounce = debounce.default;
exports.queue = queue.default;
exports.lock = lock.default;
exports.chan = chan.default;
exports.parallel = parallel.default;
exports.run = run.default;
exports.deprecate = deprecate.default;
exports.pipe = pipe.default;
exports.AsyncFunction = types.AsyncFunction;
exports.AsyncGeneratorFunction = types.AsyncGeneratorFunction;
exports.GeneratorFunction = types.GeneratorFunction;
exports.TypedArray = types.TypedArray;
exports.Channel = Channel;
exports.Mutex = Mutex;
exports.Queue = Queue;
exports.default = jsext;
exports.example = example;
exports.isClass = isClass;
exports.isSubclassOf = isSubclassOf;
exports.mixins = mixins;
exports.read = read;
exports.readAll = readAll;
